src/piadm.sh

```diff
@@ -199,6 +218,104 @@ mount_installmedia() {
 	return 0
 }
 
+# README file for /${bootfs}/platform-ipxe/README.
+cat_readme() {
+    cat <<EOF
+For iPXE boots, the platform/ directory is empty.  This README, and
+the word "ipxe" in platform/etc/version/platform, are here so there's
+something in the platform/ directory to prevent piadm (especially
+older versions) from thinking something is wrong.
+EOF
+}
+
+# Install iPXE on a pool.
+#
+# iPXE can be booted one of two ways from a disk.  If the disk has an
+# EFI System Partition (ESP), it can boot iPXE straight off of the
+# ESP.  If it is a BIOS disk, or loader is in the ESP, it will boot to
+# loader like SmartOS, but instead of using "unix" and "boot_archive"
+# from the platform directory, it will instead boot "ipxe.lkrn" and
+# "default.ipxe" from the boot directory.
+#
+# Because of the desire to have a Triton Compute Node possibly boot to
+# a recovery SmartOS, we will never install iPXE on the ESP.
+#
+# "bootfs" has been set already.
+install_ipxe() {
+	cd /${bootfs}
+	if [[ -e platform-ipxe || -e boot-ipxe ]]; then
+		eecho "existing ipxe appears to exist, version: " \
+			$(cat etc/version/ipxe)
+		err "Use 'piadm update ipxe' to upgrade it"
+	fi
+
+	# The "platform" directory for on-disk iPXE is a placeholder.
+	# We put a README (see cat_readme() above) and the string "ipxe"
+	# for the PI-stamp.
+	mkdir -p ./platform-ipxe/etc/version
+	cat_readme > ./platform-ipxe/README
+	echo "ipxe" > ./platform-ipxe/etc/version/platform
+
+	# Entry to this function is predicated by existence of
+	# TRITON_IPXE_LKRN and TRITON_IPXE_ARCHIVE.  Assume
+	# TRITON_IPXE_ETC is there too.
+	mkdir -p etc/version
+	cp -f ${TRITON_IPXE_ETC}/version/ipxe etc/version/.
+	# So 'piadm list' doesn't seize up...
+	echo "ipxe" > etc/version/boot
+	vecho "installing ipxe version: " $(cat etc/version/ipxe)
+
+	mkdir boot-ipxe
+	tar -cf - -C ${TRITON_IPXE_BOOT} . | tar -xf - -C boot-ipxe
+
+	# Populate loader.conf.
+	# XXX KEBE SAYS make loader.conf work for ipxe.
+	# XXX KEBE SAYS FILL ME IN!  AND ONCE FILLED IN CHECK ME!
+	sed 's/headnode="true"/headnode="false"/g' \
+	    < boot-ipxe/loader.conf.tmpl > boot-ipxe/loader.conf
+	echo 'ipxe="true"' >> boot-ipxe/loader.conf
+	echo 'smt_enabled="true"' >> boot-ipxe/loader.conf
```

Comment:
Looks like there's still more you plan to do here, but my initial thought is that perhaps we should leave set_enabled at whatever value it currently is. We don't want to flip this if it's currently set opposite.

Reply:
Correct, all of this is very much up in the air still (hence the KEBE comments).  My machines seem to have it set to enabled, but what do I know?

Reply:
I think the default was changed at one point, but a triton install that was set up before the default change will retain `smt_enabled=true`. It's my understanding that new triton installs will default to `false`.

But in either case, if a user has explicitly chosen a value overriding the default, we want to retain that choice and not potentially change it on them unexpectedly.

Reply:
OS-8209 may help here.  Jorge S. is working it right now, see #958 .